url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/LevelScene.java

```diff
@@ -0,0 +1,1076 @@
+            }
+        }
+
+//        for (int w = 0; w < mergedZ.length; w++)
+//            for (int h = 0; h < mergedZ[0].length; h++)
+//                mergedZ[w][h] = -1;
+//        mergedZ[Environment.HalfObsWidth][Environment.HalfObsHeight] = mario.kind;
+        for (Sprite sprite : sprites)
+        {
+            if (sprite.kind == mario.kind)
+                continue;
+            if (sprite.mapX >= 0 &&
+                sprite.mapX > MarioXInMap - Environment.HalfObsWidth &&
+                sprite.mapX < MarioXInMap + Environment.HalfObsWidth &&
+                sprite.mapY >= 0 &&
+                sprite.mapY > MarioYInMap - Environment.HalfObsHeight &&
+                sprite.mapY < MarioYInMap + Environment.HalfObsHeight )
+            {
+                int obsX = sprite.mapY - MarioYInMap + Environment.HalfObsHeight;
+                int obsY = sprite.mapX - MarioXInMap + Environment.HalfObsWidth;
+                // quick fix TODO: handle this in more general way.
+                if (mergedZ[obsX][obsY] != 14)
+                {
+                    byte tmp = ZLevelEnemyGeneralization(sprite.kind, ZLevelEnemies);
+                    if (tmp != Sprite.KIND_NONE)
+                        mergedZ[obsX][obsY] = tmp;
+                }
+            }
+        }
+
+        return mergedZ;
+    }
+
+    public List<String> LevelSceneAroundMarioASCII(boolean Enemies, boolean LevelMap,
+                                                   boolean mergedObservationFlag,
+                                                   int ZLevelScene, int ZLevelEnemies){
+//        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(os));//        bw.write("\nTotal world width = " + level.width);
+        List<String> ret = new ArrayList<String>();
+        if (level != null && mario != null)
+        {
+            ret.add("Total world width = " + level.width);
+            ret.add("Total world height = " + level.height);
+            ret.add("Physical Mario Position (x,y): (" + mario.x + "," + mario.y + ")");
+            ret.add("Mario Observation Width " + Environment.HalfObsWidth*2);
+            ret.add("Mario Observation Height " + Environment.HalfObsHeight*2);
+            ret.add("X Exit Position: " + level.xExit);
+            int MarioXInMap = (int)mario.x/16;
+            int MarioYInMap = (int)mario.y/16;
+            ret.add("Calibrated Mario Position (x,y): (" + MarioXInMap + "," + MarioYInMap + ")\n");
+
+            byte[][] levelScene = getLevelSceneObservationZ(ZLevelScene);
+            if (LevelMap)
+            {
+                ret.add("~ZLevel: Z" + ZLevelScene + " map:\n");
+                for (int x = 0; x < levelScene.length; ++x)
+                {
+                    String tmpData = "";
+                    for (int y = 0; y < levelScene[0].length; ++y)
+                        tmpData += mapElToStr(levelScene[x][y]);
+                    ret.add(tmpData);
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 516:522, 550:556.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.